### DISCUSSION TITLE
Archive viewers and viewing solutions

### DISCUSSION THREAD

**Discussion by mootari** [rusq/slackdump#127] [General]
I assume that slackdump is used for a large variety of cases, most of them being related to archival or public access.

I'd like to use this discussion to:
- get a better understanding of the use cases
- explore existing archive viewing solutions and their technical limitations
- explore workarounds, e.g. importing an archive into other chat solutions
- discuss data wrangling options (e.g. remove PII, reimport DMs as channels)

At some point some of these topics may need to be split out into separate discussions, but for the moment I'd like to keep them under one umbrella. Ultimately my hope is that I will be able to create a simple static viewer that can be bundled with an export, or a small and maintainable reference app.

**Comment by rusq** [comment]
Thanks @mootari, that's a great start.  Right now everyone has to rely on the thirdparty solution, as obviously I lack frontend skills to produce something of value in this field.  Over the time I have seen three viewers, I'll list them here for completeness:
1. [Slack Export Viewer](https://github.com/hfaran/slack-export-viewer) - Python/flask, most questions are related to this viewer.  When I adjusted the export format in the past, I aimed for being compatible with it.  Slackdump still has some compatibility issues that cause it to crash and burn
2. [SlackLogView](https://github.com/thayakawa-gh/SlackLogViewer) - C++/QT5.  Extremely fast and neat viewer, very responsive author and contributors.  It is written in C++, and currently does not support viewing of DMs, I was planning to open a couple of suggestions, namely - (a) support for DM/MPIM previews and mattermost file format support, so it would be able to preview the files straight away. 
3. [slack viewer](https://github.com/bsimpson/slack_viewer) - JS/node, learned about it recently from [this issue](https://github.com/rusq/slackdump/issues/111#issuecomment-1230645308).  It does not seem to be recently maintained, last change was around 2 years ago, according to the master branch.  Haven't tried it personally but it is reported to work with some minor changes.

**Reply by thayakawa-gh** [reply]
Hi, @rusq and @mootari. I'm the developper of SlackLogViewer and was invited here by @rusq.

Thank you for creating such a great software.
Recently I have created a branch `v1.2` to enable SlackLogViewer to read export files output by slackdump. Until now, several impediments to interoperability have been removed, and DMs/MPIMs can be displayed in SlackLogViewer. If you want to test it, you can download the binary [from here](https://github.com/thayakawa-gh/SlackLogViewer/actions/runs/3029568329).

But there are some difficulties.

- When the `-download` flag is used in slackdump, all files are downloaded in the `attachment` directories, and `url_private_download` are replaced with the relative paths to them. Is that the same behavior as the official slack export? If so, I'll allow SlackLogViewer to load them from the paths.
- To add a token string to URLs of each file, I tried the `export-token` flag on `v2.2.0-alpha`, but got an error message "flag provided but not defined: -export-token".
- Katakana characters in channel names are not properly handled. More details are [in this article](https://phst.hateblo.jp/entry/2022/08/31/210000). This is probably a problem only for Japanese users.
- I haven't tested it with official export files, as I'm just an owner of a free plan workspace.

Thank you.

**Reply by rusq** [reply]
Hi @thayakawa-gh,

Thank you for finding time to add the compatibility to the file export.
That's great news re v1.2, I'll have a look and share my experience.

1. No, the official slack export does not update the file links - it does not contain files at all, because of the following: When Slack generates an export, it generates a special "export token", which looks like other slack tokens, but with `xoxe` prefix.  So the original slack export contains links to the files on slack server, i.e.
```json
{
   "URLDownload": "https://files.slack.com/.../file.jpg?t=xoxe-*********"
}
```
With `xoxe-...` being a Slack export token generated by slack at export time.  Please have a look at [questions](#questions) section below.
2. The `-export-token` is in master, but not in the `v2.2.0-alpha` release, it will be released as a part of v2.2.0, I just need to find some time to go through the changes in master, review them and finish some tests, before v2.2.0 is out.  I will compile and attach the master for windows in the next comment to this thread.
3. Thank you for linking this, I had no experience with Katakana characters in slack export, I'll go through the article and see what changes are required to make it work.
4. I don't have enterprise access either, so been testing it with a free export as well.

## Questions
I'm keen to hear your feedback on the files support:

Currently, with the standard export option, as you have already discovered, the files are placed in the directory `attachments` and are named using the following pattern "FILE_ID-filename.ext", and the URL is updated to link to that local directory/file.  This is the solution I originally had for files since the beginning (the URL update function has been added recently, before they would go unchanged).

I can implement a special "server" mode for slackdump, that will serve all files on "http://localhost:<some_port>/files-pri/<team_id>-<FILE_ID>/download/file.jpg", so the only difference will be the hostname.  I think this would be the easiest solution to allow showing the local files.

Example:
```
# original
https://files.slack.com/files-pri/THY5HTZ8U-F03Q1KNGE3C/funny.gif
# local
https://localhost:51467/files-pri/THY5HTZ8U-F03Q1KNGE3C/funny.gif
```

The only change required from the viewer, would be to support the "local" mode of operation, so when user switches to "local" mode, the viewer would follow the following algo (draft):

1. Read the export json
2. If file is found and running in "local" mode:
  - replace the "files.slack.com" to "localhost:<some port>" in URI in URLDownload field.
3. Get the file from URL.
   
Let me know if that's something that you'd be willing to implement, and I'll create the server mode.

**Reply by rusq** [reply]
v2.2.0-alpha-53-g636e14e

[slackdump-windows.zip](https://github.com/rusq/slackdump/files/9595211/slackdump-windows.zip)

**Reply by thayakawa-gh** [reply]
Thank you for the prebuild binary. I have tested it and found out several things.

My opinions for the compatibility are the following.

1. The server mode seems to be one of the easiest way for me to implement, but in terms of usability, I am concerned that users must consciously switch the modes on SlackLogViewer if they want to load both official and slackdump's exports.

1. SlackLogViewer attempts to download attached files from the URLs in `url_private_download`, but since the URLs in slackdump's exports output without `-download` flag don't have tokens, SlackLogViewer always fails to download. That's why I tried the `-export-token` flag. If the flag appended the token to the `url_private_download` as I expected, SlackLogViewer could just read it.  
When I tried `-export-token` flag of the master, the token was appended to the URLs of sumbnail images. However, the URLs in `url_private_download` were all replaced with the relative paths to the downloaded files, although I didin't use the `-download` flag. If that was how you intended it to work, this flag would not be available for the interoperability.

1. It is also possible to have SlackLogViewer get files from the attachment directory as a special case for slackdump's exports, but it will take some time to implement, and we will need to support each other's specifications as well as the official slack export.

They all have their advantages and disadvantages, and the best way will depends on the degree of the compatibility we wish to achieve.

**Reply by rusq** [reply]
Thank you for your analysis @thayakawa-gh , I'll need to think about this.

1. I see your point, and agree that having user to constantly switch modes is not ideal.

2. If you run ./slackdump -export-type mattermost - it will not replace the `url_private_download`.  I should probably make a switch to control that behaviour for the standard export type, otherwise it seems it breaks compatibility, and I have had a number of reports for this. (For reference - updating the `url_private_download` to point to the file was done as a part of https://github.com/rusq/slackdump/discussions/70)

**Reply by rusq** [reply]
@thayakawa-gh the issue with folder naming is fixed in [v2.2.1](https://github.com/rusq/slackdump/releases/tag/v2.2.1).  Thanks for flagging it!

I'm sorry I still haven't checked the SlackLogViewer, but I'll let you know when I do!

**Reply by rusq** [reply]
@thayakawa-gh playing around with SlackLogViewer — it works great! even shows the reaction emojis.

- Threads are working
- Files are expected not to work for now, but they are shown
- The DMs are displayed correctly, search within DMs works
- All UI interface elements are working on macOS without any adventures :)

Overall - great work! I think if you're planning to release it, it's good to go - I'll link to SlackLogViewer from the main README.  btw, I really enjoy the "All channels" search function  - it turns an export file into a searchable knowledge base.  And also, the overall speed of the SlackLogViewer is amazing, comparing to the modern web-apps, which inevitably lagging due to their nature.

I still need to think about the approach which we discussed earlier, will share my thoughts, once I come up with something.

**Reply by thayakawa-gh** [reply]
Many thanks for fixing the issue and testing with SlackLogViewer.
I'd like to test v2.2.1 too, but I have no time to do so at all for a while. Sorry, but please wait a week or so. I'll also release the new version of SlackLogViewer at that time.

**Reply by thayakawa-gh** [reply]
I tested slackdump v2.2.2 with my workspace including some Katakana channel names and confirmed that they were correctly dumped.
Excellent! Thank you for fixing, this would be a great help for all Japanese users.

I also confirmed that files in exports dumped with `-export-type mattermost` and `-export-token` could be shown in SlackLogViewer.

I have released v1.2.Alpha-0 including a minor bug fix. It would be helpful if you could change the link from the actions to the top or release page.

**Reply by rusq** [reply]
Hi @thayakawa-gh,

I'm glad that it works and that it will make life easier for Japanese users!  I have linked the releases page in the readme and docs, so that users would get the latest version (merged to master, see https://github.com/rusq/slackdump/pull/161).

Thank you for testing and your feedback, greatly appreciated!

**Comment by mootari** [comment]
### Data migration

- Rocket.Chat can import corporate exports that include DMs (more info [here](https://github.com/RocketChat/Rocket.Chat/issues/13905) or in the PR referenced at the end).
- Mattermost can import corporate Slack exports directly as well ([issue](https://github.com/mattermost/mattermost-server/pull/11704))
- Discord: There are contrib solutions (e.g. [Slackord](https://github.com/thomasloupe/Slackord), [slack2discord](https://github.com/rslavin/slack2discord))

**Reply by rusq** [reply]
I did some digging on the mattermost in scope of #107, there's a bulk import format.  As the structures are open source, and part of mattermost-server, which [looks compatible with GPL 3][3], so at some point I was planning to use that structures to add the support for direct export to Mattermost bulk format.  But I'm still undecided whether I should leave this part to the user, and just limit the scope of Slackdump to slack only, to avoid the burden of supporting another format in the future, especially that there's a [mmetl][1] tool that does the job, with some [quirks][2] though requiring manual effort of shuffling files around, which could be a quick fix to apply to the mmetl tool itself, but i got the impression that the support of that tool is not a priority for the mattermost team, looking at the size of the mattermost project, they have plenty other things to take care of :)

[1]: https://github.com/mattermost/mmetl
[2]: https://github.com/mattermost/mattermost-server/issues/19747#issue-1163102268
[3]: https://github.com/mattermost/mattermost-server/blob/master/LICENSE.txt

**Reply by mootari** [reply]
I'm not familiar with either format, but mmetl seems to take the right approach. I would probably recommend to focus on Slack's export format (which would likely be enough work in itself, given that more features might be added to Slack in the future), and contribute to mmetl if necessary.

It may also be worth asking who's going to benefit from a direct export to Mattermost from slackdump. I would expect such a conversion to be a one-time step in a migration from Slack to Mattermost, not a repeatedly performed action. Conversion guides and scripts could be documented in this repo's wiki and (hopefully) be maintained by the community.

**Reply by mootari** [reply]
... In the long run it would be awesome to have some automated testing capabilities:
1. A workspace with some test data (archived channels?) and a couple of test users
2. A reference export that is used to spot any changes in the export structure.
3. Puppeteer based export perhaps, and import into other tools like e.g. Mattermost.

This is admittedly very hand wavy and probably a lot (like, *a lot*) more complicated than I make it sound here.

**Reply by rusq** [reply]
@mootari very good point on testing!  Right now there's no integrational tests, only unit, and coverage is less than I'd like it to be, but more than the threshold that was giving me the itch 😂

**Comment by mootari** [comment]
### Browsing UIs

- [linen](https://www.linen.dev/) exposes workspaces to Google. It has channel selection and thread displays ([example](https://slack-chats.kotlinlang.org/)).
- [slack-archive](https://github.com/felixrieseberg/slack-archive) exports messages and makes them accessible in a static UI (apparently? haven't verified yet). Seems we should be able to replace [`data-load.js`](https://github.com/felixrieseberg/slack-archive/blob/main/src/data-load.ts) with an adapter for slackdump exports (if the file format doesn't already match). Most promising solution so far.
- ~[Block Kit](https://api.slack.com/block-kit) is Slack's own style guide and component library for Slack apps.~

Considerations:
- Display of DMs could emulate the Slack UI.
- In theory all data could be made available as static JSON, but that may require an insane amount of data to download and keep in memory. Alternative storage formats like Apache Arrow might be more suitable here.

**Reply by mootari** [reply]
I've blocked some time on Friday to look into slack-archive. Curious to see how "static" their static HTML is. At first glance it seems they use React at least in an intermediate step (maybe along with some SSR).

**Reply by mootari** [reply]
Block Kit may be a better fit to handle Slack's message storage format. I'll investigate that one first.

**Reply by mootari** [reply]
Sadly it seems that Block Kit is mostly a proprietary API for Slack apps that does not include all UI elements required to recreate the entire Slack UI.

**Reply by mootari** [reply]
I don't think I want to use any module from slack-archive as-is. The code is heavily interleaved with console/progress stuff. I'll probably extract or port their components instead. They reference interfaces from `@slack/wep-api`, so I expect compatibility with slackdump's export to be high.

**Comment by mazerj** [comment]
I was just playing around with slack-export-viewer (flask-based solution #1 above) and noticed that it mostly works, but looks for attachments in the wrong location. Or rather, slackdump puts them in a different directory than the viewer expects them to be. Not sure if this is a change in the format, but figured I'd ask before I tried to fix or patch a solution.
To be specific. slack-export-viewer is looking for attachments associated with DMs in: `http://localhost:8080/dm/<ID>/attachments/FKW3PMKNH-mlab-bulb2.png`
and (as far as I can tell) the correct URL should be: `http://localhost:8080/<ID>/attachments/FKW3PMKNH-mlab-bulb2.png` (that is w/o the `dm` subdir.
Not sure this whether this is a slack-export-viewer or slackdump issue. Any thoughts before I try to change something?

**Comment by rusq** [comment]
Slackdump has `slackdump view` command that allows to view archives, since v3.0.

**Reply by volker-fr** [reply]
There are still benefits from slack-export-viewer. It actually has two features: web server & export to a html. The export allows templating. As result I can export every week all new posts, including showing new messages in old threads.

**Reply by rusq** [reply]
There's no doubt, I quite like it myself.  To me the benefit of having a built in viewer is to have a control over features, bugs... When someone creates an issue that X can't view images, I can always point to built-in viewer that can. None of the external viewers support the database format, and I doubt anyone would be interested or have time to add it. But for native exports these tools are indispensable.
